### PR TITLE
[C/PyTorch] Add support for bottom-right-diagonal causal mask

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

Currently, our causal masks are aligned to the top left corner of the softmax matrix, but in inference/KV caching, users often need to align them to the bottom right corner. This PR adds two mask types, `causal_bottom_right` and `padding_causal_bottom_right`, to support this new alignment. The old mask types, `causal` and `padding_causal`, are still used to denote the top-left alignment.

The new support matrix is,
```
attn_mask_type               |     supported backends
-------------------------------------------------------------------
no_mask                      |     All  
padding                      |     FlashAttention, FusedAttention
causal                       |    
    self-attention           |     All  
    cross-attention          |     FusedAttention
padding_causal               |    
    self-attention           |     FlashAttention, FusedAttention
    cross-attention          |     FusedAttention
causal_bottom_right          |     All  
padding_causal_bottom_right  |     FlashAttention, FusedAttention
arbitrary                    |     UnfusedDotProductAttention
```

This PR also extracts and streamlines the utility function `get_attention_backend()` for backend availability test. Users can call it with their model params/runtime environment to check which backends are available to support a particular set of user input, and which backend will be selected based on TransformerEngine's internal logic.

To facilitate the addition of bottom-right causal masks, this PR also makes two other changes for decoders (PR #895 ).
- improves `check_set_window_size` function to make `window_size` consistent with `attn_mask_type`
- adds `enc_dec_attn_mask_type` and `enc_dec_window_size` parameters to `TransformerLayer` so encoder MHA call and decoder MHA call are separated.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Add bottom-right-diagonal support for causal mask on both C and PyTorch levels
- Add utility function `get_attention_backend` to help with availability tests
- Add mask type and window size params for decoder situations

# Checklist:

- [x ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x ] The functionality is complete
- [x ] I have commented my code, particularly in hard-to-understand areas
- [x ] I have made corresponding changes to the documentation
- [x ] My changes generate no new warnings
- [x ] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 1]
```diff
@@ -177,11 +177,11 @@ def _is_flash_attention_supported(config: ModelConfig) -> bool:
...
-    if "causal" in config.attn_mask_type and config.attn_type == "cross":
-        if _is_flash_attention_2_1():
-            # FAv2.1 implements causal mask for cross attention differently
-            # https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag
-            return False
+    # if "causal_bottom_right" in config.attn_mask_type and config.attn_type == "cross":
+    #    if _is_flash_attention_2_1():
+    #        # FAv2.1 implements causal mask for cross attention differently
+    #        # https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag
+    #        return False
```
Comment (timmoon10) [inline tests/pytorch/fused_attn/test_fused_attn.py:184]: Debugging code?
```suggestion
    if "causal" in config.attn_mask_type and config.attn_type == "cross":
       if _is_flash_attention_2_1():
           # FAv2.1 implements causal mask for cross attention differently
           # https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag
           return False
```
Also, do we need a similar check to disable bottom-right causal for FA <2.1?

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -4962,17 +4985,15 @@ def forward(
...
                 and max_seqlen_q != max_seqlen_kv
             ):
                 self.logger.warning(
-                    "In training mode, disable the use of FlashAttention since version 2.1+ has "
-                    "changed its behavior for causal mask in cross attention. See "
+                    "Disabling FlashAttention as it only supports bottom-right-diagonal "
+                    "causal mask since version 2.1. See "
                     "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
                 )
                 use_flash_attention = False
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:4999]: Do we need a similiar check to disable bottom-right causal masks for FA <2.1?

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 3]
```diff
@@ -381,18 +381,20 @@ def test_dpa_checkpoint(dtype, model_configs, model):
...
+    # "mask_3_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "padding", "no_bias"),
+    # "mask_3_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "padding", "no_bias"),
+    # "mask_4_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "padding", "no_bias"),
+    # "mask_4_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "padding", "no_bias"),
+    # "mask_5_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "padding_causal", "no_bias"),
+    # "mask_5_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "padding_causal", "no_bias"),
+    # "mask_6_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "padding_causal", "no_bias"),
+    # "mask_6_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "padding_causal", "no_bias"),
+    "mask_2_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "causal_bottom_right", "no_bias"),
+    "mask_2_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "causal_bottom_right", "no_bias"),
```
Comment (timmoon10) [inline tests/pytorch/fused_attn/test_fused_attn.py:397]: Debugging code?
```suggestion
    "mask_1_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "causal", "no_bias"),
    "mask_1_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "causal", "no_bias"),
    "mask_2_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "causal", "no_bias"),
    "mask_2_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "causal", "no_bias"),
    "mask_3_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "padding", "no_bias"),
    "mask_3_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "padding", "no_bias"),
    "mask_4_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "padding", "no_bias"),
    "mask_4_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "padding", "no_bias"),
    "mask_5_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "padding_causal", "no_bias"),
    "mask_5_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "padding_causal", "no_bias"),
    "mask_6_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "padding_causal", "no_bias"),
    "mask_6_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "padding_causal", "no_bias"),
    "mask_7_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "causal_bottom_right", "no_bias"),
    "mask_7_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "causal_bottom_right", "no_bias"),
```

File: tests/pytorch/fused_attn/test_fused_attn.py (hunk 1 above)
Comment (cyanguwa) [reply tests/pytorch/fused_attn/test_fused_attn.py]: Ya, a lot of debugging code in that commit.

Code is refactored, and a check for FA 2.1 is add [here](https://github.com/cyanguwa/TransformerEngine/blob/4d2c55a1ae087e5c7409c08701d66a2eaff9af0f/transformer_engine/pytorch/attention.py#L352).

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -116,6 +117,416 @@
...
+        and _flash_attn_2_1_plus
+        and attn_mask_type in ["causal", "padding_causal"]
+        and max_seqlen_q != max_seqlen_kv
+    ):
+        logger.warning(
+            "Disabling FlashAttention as it only supports bottom-right-diagonal "
+            "causal mask since flash-attn 2.1. See "
+            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
+        )
+        use_flash_attention = False
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:353]:
```suggestion
    if (
        use_flash_attention
        and _flash_attn_2_1_plus
        and attn_mask_type in ["causal", "padding_causal"]
        and max_seqlen_q != max_seqlen_kv
    ):
        logger.warning(
            "Disabling FlashAttention as it only supports bottom-right-diagonal "
            "causal mask since flash-attn 2.1. See "
            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
        )
        use_flash_attention = False
    if (
        use_flash_attention
        and not _flash_attn_2_1_plus
        and attn_mask_type in ["causal_bottom_right", "padding_causal_bottom_right"]
        and max_seqlen_q != max_seqlen_kv
    ):
        logger.warning(
            "Disabling FlashAttention as it only supports top-left-diagonal "
            "causal mask before flash-attn 2.1. See "
            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
        )
        use_flash_attention = False
```
